lib/mllp/client.ex

```diff
@@ -464,6 +470,74 @@ defmodule MLLP.Client do
     stop_connection(state, reason, "process terminated")
   end
 
+  defp recv_ack(state, :infinity), do: recv_ack(state, :infinity, 0, <<>>)
+
+  defp recv_ack(state, timeout) do
+    recv_ack(state, timeout, 0, <<>>)
+  end
+
+  defp recv_ack(_state, timeout, time_spent, _buffer)
+       when is_integer(timeout) and timeout * 1000 <= time_spent do
+    {:error, :timeout}
+  end
+
+  defp recv_ack(state, timeout, time_spent, buffer) do
+    {res, elapsed} = do_recv(state, 0, timeout)
+
+    case res do
+      {:ok, reply} ->
+        new_buf = buffer <> reply
+        check = byte_size(new_buf) - 3
+
+        case new_buf do
+          <<@header, _ack::binary-size(check), @trailer>> ->
+            {:ok, new_buf}
+
+          <<@header, _rest::binary>> ->
+            recv_ack(
+              state,
+              maybe_decrement_timeout(timeout, time_spent + elapsed),
```

Comment:
Had to adjust how we're doing this. If a client supplies a 5 millisecond timeout (as an example), and the ack/nack comes in 6 segments; in the previous implementation we would have bailed after about 20 microseconds. 

This version should be more accurate.

Reply:
I am wondering, if we should keep `reply_timeout` in `microsecond` for calculation, and only convert it before `:recv` call. Wouldn't this make your calculation more simple? 
`case recv_ack(state, options1.reply_timeout) do` call should be `case recv_ack(state, System.convert_time_unit(options1.reply_timeout), :millisecond, :microsecond)) do`

Reply:
I think I'm following you 🤔  I'm not sure which part it would simplify though. Maybe it's the ` timeout - div(spent, 1000)` ? 

If so, we'd have to do that anyway, the reason exemplified is : 

```
iex(1)> micros = System.convert_time_unit(5000, :millisecond, :microsecond)
5000000
iex(2)> micros - 1001
4998999
iex(3)> updated = micros - 1001
4998999
iex(4)> System.convert_time_unit(updated, :microsecond, :millisecond)
4998
```

So, we lose 2 ms vs 1, that big of decrement in the case of small reply_timeout values is a big difference I believe.

But I may be off here and you might be pointing at a simplification I'm not seeing or thinking of.

Reply:
This is the simplification, I was talking about. I don't seem to loose time in it. Every time we try to recv and get a response, we reduce our timeout in microsecond from elapsed time. if timeout becomes zero anytime, we timeout otherwise carry on.

```
  defp recv_ack(state, timeout) do
    recv_ack(state, timeout, <<>>)
  end

  defp recv_ack(_state, time_left, _buffer)
       when is_integer(time_left) and time_left <= 0 do
    {:error, :timeout}
  end

  defp recv_ack(state, time_left, buffer) do
    {res, elapsed} = do_recv(state, 0, time_left)

    case res do
      {:ok, reply} ->
        new_buf = buffer <> reply
        check = byte_size(new_buf) - 3

        case new_buf do
          <<@header, _ack::binary-size(check), @trailer>> ->
            {:ok, new_buf}

          <<@header, _rest::binary>> ->
            recv_ack(state, time_left - elapsed, new_buf)

          _ ->
            {:error, :invalid_reply}
        end

      {:error, _} = err ->
        err
    end
  end
```
```
iex(8)> micros = System.convert_time_unit(5000, :millisecond, :microsecond)
5000000
iex(9)> elapsed = 1000                                                     
1000
iex(10)> updated = micros - elapsed                                         
4999000
iex(11)> System.convert_time_unit(updated, :microsecond, :millisecond)      
4999
```

Reply:
how about if we calculate elapsed time in `millisecond` itself. Worst case, we will over weight a bit but would never bail out early